File: apps/gate/src/pipeline.ts
// The Action Gate pipeline (DESIGN.md section 4):
//   capability check -> approval check -> run budget -> idempotency -> audit + cost log -> execute
// Every external effect in the system goes through requestAction/approveAction.
import type pg from "pg";
import { pool } from "./db.js";
import { payloadHash } from "./hash.js";
import { CAPABILITIES, REQUIRES_ADMIN_APPROVAL } from "./capabilities.js";

export interface ActionRequest {
  tenantId: string;
  runId?: string | null;
  agentName: string;
  actionType: string;
  payload: unknown;
  idempotencyKey: string;
}

export interface ExecutorResult {
  providerReference: string;
  providerCostMicrodollars?: number;
  output?: unknown;
}

export type Executor = (
  payload: unknown,
  ctx: { tenantId: string; actionId: string; mode: string },
) => Promise<ExecutorResult>;

export interface PipelineConfig {
  executors: Record<string, Executor>;
  capabilities?: Readonly<Record<string, readonly string[]>>;
  requiresApproval?: ReadonlySet<string>;
}

export class GateError extends Error {
  constructor(
    public readonly httpStatus: number,
    message: string,
  ) {
    super(message);
  }
}

export interface ActionRow {
  id: string;
  tenant_id: string;
  run_id: string | null;
  agent_name: string;
  action_type: string;
  payload_hash: string;
  idempotency_key: string;
  mode: string;
  approval_status: string;
  approved_by: string | null;
  approved_at: Date | null;
  provider_reference: string | null;
  provider_cost_microdollars: string;
  status: string;
  created_at: Date;
  executed_at: Date | null;
}

const MODE = (): string => process.env.RUN_MODE ?? "TEST";

async function runSpendMicrodollars(client: pg.PoolClient, runId: string): Promise<bigint> {
  const { rows } = await client.query<{ spent: string }>(
    `SELECT COALESCE((SELECT SUM(cost_microdollars) FROM agent_calls WHERE run_id = $1), 0)
          + COALESCE((SELECT SUM(provider_cost_microdollars) FROM actions WHERE run_id = $1), 0)
          AS spent`,
    [runId],
  );
  return BigInt(rows[0]?.spent ?? "0");
}

async function execute(
  row: ActionRow,
  payload: unknown,
  config: PipelineConfig,
): Promise<ActionRow> {
  const executor = config.executors[row.action_type];
  if (!executor) {
    throw new GateError(501, `no executor implemented for action type '${row.action_type}'`);
  }
  try {
    const result = await executor(payload, {
      tenantId: row.tenant_id,
      actionId: row.id,
      mode: row.mode,
    });
    const { rows } = await pool.query<ActionRow>(
      `UPDATE actions
         SET status = 'executed',
             provider_reference = $2,
             provider_cost_microdollars = $3,
             executed_at = now()
       WHERE id = $1
       RETURNING *`,
      [row.id, result.providerReference, result.providerCostMicrodollars ?? 0],
    );
    return rows[0]!;
  } catch (err) {
    await pool.query(`UPDATE actions SET status = 'failed', executed_at = now() WHERE id = $1`, [
      row.id,
    ]);
    console.error(`[gate] action ${row.id} (${row.action_type}) failed:`, err);
    throw new GateError(502, `action execution failed: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/**
 * Request an action through the gate. Returns the audit row; when approval is
 * required the row comes back status=pending_approval and nothing executes.
 * Replaying the same (tenant, action_type, idempotency_key) never executes twice.
 */
export async function requestAction(
  req: ActionRequest,
  config: PipelineConfig,
): Promise<{ row: ActionRow; replayed: boolean }> {
  const capabilities = config.capabilities ?? CAPABILITIES;
  const requiresApproval = config.requiresApproval ?? REQUIRES_ADMIN_APPROVAL;

  // 1. Capability: may this agent request this action at all?
  const allowed = capabilities[req.agentName];
  if (!allowed || !allowed.includes(req.actionType)) {
    throw new GateError(403, `agent '${req.agentName}' has no capability '${req.actionType}'`);
  }

  // 2. Mode: TEST by default; LIVE is rejected until explicitly supported.
  const mode = MODE();
  if (mode !== "TEST") {
    throw new GateError(400, `unsupported mode '${mode}' - the gate runs TEST-mode only`);
  }

  const hash = payloadHash(req.payload);
  const needsApproval = requiresApproval.has(req.actionType);

  const client = await pool.connect();
  let row: ActionRow;
  let replay = false;
  try {
    await client.query("BEGIN");

    // 3. Idempotency: one row per (tenant, action_type, key), locked.
    const existing = await client.query<ActionRow>(
      `SELECT * FROM actions
        WHERE tenant_id = $1 AND action_type = $2 AND idempotency_key = $3
        FOR UPDATE`,
      [req.tenantId, req.actionType, req.idempotencyKey],
    );

    if (existing.rows[0]) {
      row = existing.rows[0];
      if (row.payload_hash !== hash) {
        throw new GateError(
          409,
          `idempotency key '${req.idempotencyKey}' was already used with a different payload`,
        );
      }
      // executed -> replay result; pending_approval -> still waiting; both
      // short-circuit. approved/failed fall through to (re-)execution below.
      if (row.status === "executed" || row.status === "pending_approval") {
        await client.query("COMMIT");
        return { row, replayed: true };
      }
      replay = true;
    } else {
      // 4. Budget: when the action belongs to a run, enforce the approved cap.
      if (req.runId) {
        const budget = await client.query<{ approved_budget_microdollars: string }>(
          `SELECT approved_budget_microdollars FROM runs WHERE id = $1`,
          [req.runId],
        );
        if (!budget.rows[0]) throw new GateError(404, `run ${req.runId} not found`);
        const spent = await runSpendMicrodollars(client, req.runId);
        if (spent >= BigInt(budget.rows[0].approved_budget_microdollars)) {
          throw new GateError(402, `run ${req.runId} has exhausted its approved budget`);
        }
      }

      // 5. Audit row - created before execution, so even a crash leaves a trace.
      const inserted = await client.query<ActionRow>(
        `INSERT INTO actions
           (tenant_id, run_id, agent_name, action_type, payload_hash, idempotency_key,
            mode, approval_status, status)
         VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)
         RETURNING *`,
        [
          req.tenantId,
          req.runId ?? null,
          req.agentName,
          req.actionType,
          hash,
          req.idempotencyKey,
          mode,
          needsApproval ? "PENDING" : "NOT_REQUIRED",
          needsApproval ? "pending_approval" : "approved",
        ],
      );
      row = inserted.rows[0]!;
    }

    await client.query("COMMIT");
  } catch (err) {
    await client.query("ROLLBACK");
    throw err;
  } finally {
    client.release();
  }

  // 6. Execute immediately only when no human approval is needed.
  if (row.status === "pending_approval") return { row, replayed: false };
  const executed = await execute(row, req.payload, config);
  return { row: executed, replayed: replay };
}

/**
 * Approve a pending action. The approval is bound to the payload hash the
 * approver saw - a mismatch (superseded payload) is rejected outright.
 * The caller must supply the original payload so the executor can run.
 */
export async function approveAction(
  actionId: string,
  approvedBy: string,
  approvedPayloadHash: string,
  payload: unknown,
  config: PipelineConfig,
): Promise<ActionRow> {
  const client = await pool.connect();
  let row: ActionRow;
  try {
    await client.query("BEGIN");
    const { rows } = await client.query<ActionRow>(
      `SELECT * FROM actions WHERE id = $1 FOR UPDATE`,
      [actionId],
    );
    if (!rows[0]) throw new GateError(404, `action ${actionId} not found`);
    row = rows[0];

    if (row.status !== "pending_approval") {
      throw new GateError(409, `action ${actionId} is '${row.status}', not pending_approval`);
    }
    if (row.payload_hash !== approvedPayloadHash) {
      throw new GateError(
        409,
        "approval hash does not match the action's payload - the payload was superseded; re-review it",
      );
    }
    if (payloadHash(payload) !== row.payload_hash) {
      throw new GateError(409, "supplied payload does not match the approved hash");
    }

    const updated = await client.query<ActionRow>(
      `UPDATE actions
         SET approval_status = 'APPROVED', approved_by = $2, approved_at = now(),
             status = 'approved'
       WHERE id = $1
       RETURNING *`,
      [actionId, approvedBy],
    );
    row = updated.rows[0]!;
    await client.query("COMMIT");
  } catch (err) {
    await client.query("ROLLBACK");
    throw err;
  } finally {
    client.release();
  }

  return execute(row, payload, config);
}

export async function getAction(actionId: string): Promise<ActionRow | null> {
  const { rows } = await pool.query<ActionRow>(`SELECT * FROM actions WHERE id = $1`, [actionId]);
  return rows[0] ?? null;
}

export async function pendingApprovals(): Promise<ActionRow[]> {
  const { rows } = await pool.query<ActionRow>(
    `SELECT * FROM actions WHERE status = 'pending_approval' ORDER BY created_at`,
  );
  return rows;
}

File: apps/gate/gate/pipeline.py
"""The Action Gate pipeline (DESIGN.md section 4):

    capability check -> approval check -> run budget -> idempotency
    -> audit + cost log -> execute

Every external effect in the system goes through request_action/approve_action.
"""

import logging
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
from typing import Any

from .capabilities import CAPABILITIES, REQUIRES_ADMIN_APPROVAL
from .db import pool
from .hashing import payload_hash

log = logging.getLogger("gate.pipeline")

Executor = Callable[[Any, "ExecutorContext"], "ExecutorResult"]


class GateError(Exception):
    def __init__(self, http_status: int, message: str) -> None:
        super().__init__(message)
        self.http_status = http_status
        self.message = message


@dataclass
class ExecutorContext:
    tenant_id: str
    action_id: str
    mode: str


@dataclass
class ExecutorResult:
    provider_reference: str
    provider_cost_microdollars: int = 0


@dataclass
class PipelineConfig:
    executors: Mapping[str, Executor]
    capabilities: Mapping[str, tuple[str, ...]] = field(default_factory=lambda: CAPABILITIES)
    requires_approval: frozenset[str] = field(default_factory=lambda: REQUIRES_ADMIN_APPROVAL)


def _mode() -> str:
    import os

    return os.environ.get("RUN_MODE", "TEST")


def _run_spend_microdollars(cur: Any, run_id: str) -> int:
    cur.execute(
        """SELECT COALESCE((SELECT SUM(cost_microdollars) FROM agent_calls WHERE run_id = %s), 0)
                + COALESCE((SELECT SUM(provider_cost_microdollars) FROM actions WHERE run_id = %s), 0)
                AS spent""",
        (run_id, run_id),
    )
    return int(cur.fetchone()["spent"])


def _execute(row: dict[str, Any], payload: Any, config: PipelineConfig) -> dict[str, Any]:
    executor = config.executors.get(row["action_type"])
    if executor is None:
        raise GateError(501, f"no executor implemented for action type '{row['action_type']}'")
    try:
        result = executor(
            payload,
            ExecutorContext(
                tenant_id=str(row["tenant_id"]), action_id=str(row["id"]), mode=row["mode"]
            ),
        )
    except GateError:
        _mark_failed(row)
        raise
    except Exception as err:
        _mark_failed(row)
        log.exception("action %s (%s) failed", row["id"], row["action_type"])
        raise GateError(502, f"action execution failed: {err}") from err

    with pool.connection() as conn:
        updated = conn.execute(
            """UPDATE actions
                  SET status = 'executed', provider_reference = %s,
                      provider_cost_microdollars = %s, executed_at = now()
                WHERE id = %s RETURNING *""",
            (result.provider_reference, result.provider_cost_microdollars, row["id"]),
        ).fetchone()
    return updated


def _mark_failed(row: dict[str, Any]) -> None:
    with pool.connection() as conn:
        conn.execute(
            "UPDATE actions SET status = 'failed', executed_at = now() WHERE id = %s",
            (row["id"],),
        )


def request_action(
    *,
    tenant_id: str,
    agent_name: str,
    action_type: str,
    payload: Any,
    idempotency_key: str,
    run_id: str | None = None,
    config: PipelineConfig,
) -> tuple[dict[str, Any], bool]:
    """Returns (audit_row, replayed). When approval is required the row comes
    back status=pending_approval and nothing executes. Replaying the same
    (tenant, action_type, idempotency_key) never executes twice."""

    # 1. Capability: may this agent request this action at all?
    allowed = config.capabilities.get(agent_name)
    if not allowed or action_type not in allowed:
        raise GateError(403, f"agent '{agent_name}' has no capability '{action_type}'")

    # 2. Mode: TEST by default; LIVE is rejected until explicitly supported.
    mode = _mode()
    if mode != "TEST":
        raise GateError(400, f"unsupported mode '{mode}' - the gate runs TEST-mode only")

    digest = payload_hash(payload)
    needs_approval = action_type in config.requires_approval
    replay = False

    with pool.connection() as conn, conn.cursor() as cur:
        # 3. Idempotency: one row per (tenant, action_type, key), locked.
        cur.execute(
            """SELECT * FROM actions
                WHERE tenant_id = %s AND action_type = %s AND idempotency_key = %s
                FOR UPDATE""",
            (tenant_id, action_type, idempotency_key),
        )
        row = cur.fetchone()

        if row is not None:
            if row["payload_hash"] != digest:
                raise GateError(
                    409,
                    f"idempotency key '{idempotency_key}' was already used with a different payload",
                )
            # executed -> replay result; pending_approval -> still waiting;
            # both short-circuit. approved/failed fall through to (re-)execution.
            if row["status"] in ("executed", "pending_approval"):
                return row, True
            replay = True
        else:
            # 4. Budget: when the action belongs to a run, enforce the cap.
            if run_id:
                cur.execute(
                    "SELECT approved_budget_microdollars FROM runs WHERE id = %s", (run_id,)
                )
                budget_row = cur.fetchone()
                if budget_row is None:
                    raise GateError(404, f"run {run_id} not found")
                spent = _run_spend_microdollars(cur, run_id)
                if spent >= int(budget_row["approved_budget_microdollars"]):
                    raise GateError(402, f"run {run_id} has exhausted its approved budget")

            # 5. Audit row - created before execution, so a crash leaves a trace.
            cur.execute(
                """INSERT INTO actions
                     (tenant_id, run_id, agent_name, action_type, payload_hash,
                      idempotency_key, mode, approval_status, status)
                   VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s)
                   RETURNING *""",
                (
                    tenant_id,
                    run_id,
                    agent_name,
                    action_type,
                    digest,
                    idempotency_key,
                    mode,
                    "PENDING" if needs_approval else "NOT_REQUIRED",
                    "pending_approval" if needs_approval else "approved",
                ),
            )
            row = cur.fetchone()

    # 6. Execute immediately only when no human approval is needed.
    if row["status"] == "pending_approval":
        return row, False
    return _execute(row, payload, config), replay


def approve_action(
    action_id: str,
    approved_by: str,
    approved_payload_hash: str,
    payload: Any,
    config: PipelineConfig,
) -> dict[str, Any]:
    """Approve a pending action. The approval is bound to the payload hash the
    approver saw - a mismatch (superseded payload) is rejected outright."""
    with pool.connection() as conn, conn.cursor() as cur:
        cur.execute("SELECT * FROM actions WHERE id = %s FOR UPDATE", (action_id,))
        row = cur.fetchone()
        if row is None:
            raise GateError(404, f"action {action_id} not found")
        if row["status"] != "pending_approval":
            raise GateError(409, f"action {action_id} is '{row['status']}', not pending_approval")
        if row["payload_hash"] != approved_payload_hash:
            raise GateError(
                409,
                "approval hash does not match the action's payload - "
                "the payload was superseded; re-review it",
            )
        if payload_hash(payload) != row["payload_hash"]:
            raise GateError(409, "supplied payload does not match the approved hash")

        cur.execute(
            """UPDATE actions
                  SET approval_status = 'APPROVED', approved_by = %s, approved_at = now(),
                      status = 'approved'
                WHERE id = %s RETURNING *""",
            (approved_by, action_id),
        )
        row = cur.fetchone()

    return _execute(row, payload, config)


def get_action(action_id: str) -> dict[str, Any] | None:
    with pool.connection() as conn:
        return conn.execute("SELECT * FROM actions WHERE id = %s", (action_id,)).fetchone()


def pending_approvals() -> list[dict[str, Any]]:
    with pool.connection() as conn:
        return conn.execute(
            "SELECT * FROM actions WHERE status = 'pending_approval' ORDER BY created_at"
        ).fetchall()
